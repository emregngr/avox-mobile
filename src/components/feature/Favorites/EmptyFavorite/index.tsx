import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useMemo } from 'react'
import { View } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface EmptyFavoriteProps {
  icon: keyof typeof MaterialCommunityIcons.glyphMap
  text: string
}

export const EmptyFavorite = ({ icon, text }: EmptyFavoriteProps) => {
  const { bottom } = useSafeAreaInsets()

  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  return (
    <View
      className="flex-1 items-center justify-center mt-[52px] px-4"
      style={{ marginBottom: bottom + 60 }}
      testID="EmptyFavoriteContainer"
    >
      <View className="bg-background-secondary p-6 rounded-full overflow-hidden mb-6">
        <MaterialCommunityIcons color={colors?.onPrimary100} name={icon} size={64} />
      </View>

      <ThemedText color="text-100" type="h1" center>
        {text}
      </ThemedText>
    </View>
  )
}
