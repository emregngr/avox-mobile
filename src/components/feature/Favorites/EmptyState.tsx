import React from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common'

interface EmptyStateProps {
  text: string
}

export const EmptyState = ({ text }: EmptyStateProps) => (
  <View className="flex-1 items-center justify-center px-4 min-h-full">
    <ThemedText color="text-100" type="h1">
      {text}
    </ThemedText>
  </View>
)
