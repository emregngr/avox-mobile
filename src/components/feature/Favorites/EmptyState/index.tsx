import React from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'

interface EmptyStateProps {
  text: string
}

export const EmptyState = ({ text }: EmptyStateProps) => (
  <View className="flex-1 items-center justify-center px-4 min-h-full">
    <ThemedText color="text-100" type="h1" center>
      {text}
    </ThemedText>
  </View>
)
