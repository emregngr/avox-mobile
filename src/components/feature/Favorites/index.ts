export * from './FavoriteAirlinesList'
export * from './FavoriteAirportsList'
