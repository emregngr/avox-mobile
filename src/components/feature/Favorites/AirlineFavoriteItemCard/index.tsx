import { Image } from 'expo-image'
import { router } from 'expo-router'
import React, { memo, useCallback, useMemo } from 'react'
import { Platform, TouchableOpacity, View } from 'react-native'
import { TestIds } from 'react-native-google-mobile-ads'

import { ThemedText } from '@/components/common/ThemedText'
import { FavoriteButton } from '@/components/feature/FavoriteButton'
import { useInterstitialAdHandler } from '@/hooks/advertisement/useInterstitialAdHandler'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import type { Airline } from '@/types/feature/airline'
import { AnalyticsService } from '@/utils/common/analyticsService'
import { responsive } from '@/utils/common/responsive'
import type { AirlineType } from '@/utils/feature/getBadge'
import { getAirlineBadge } from '@/utils/feature/getBadge'

interface AirlineFavoriteItemCardProps {
  airline: Airline
}

const AD_UNIT_ID = __DEV__
  ? TestIds.INTERSTITIAL
  : Platform.OS === 'ios'
    ? 'ca-app-pub-4123130377375974/1756124081'
    : 'ca-app-pub-4123130377375974/2330839152'

const STATIC_STYLES = {
  badge: {
    bottom: 4,
    height: 40,
    left: 4,
    position: 'absolute' as const,
    width: 40,
  },
  containerWidth: {
    width: responsive.deviceWidth / 2 - 44,
  },
  title: {
    height: 130,
  },
}

const AirlineFavoriteItemCard = memo(({ airline }: AirlineFavoriteItemCardProps) => {
  const { selectedLocale } = useLocaleStore()

  const {
    fleet: { totalAirplane },
    iataCode,
    icaoCode,
    id,
    name,
    network: { destinationCount, destinationCountries, destinations },
    operations: {
      businessModel,
      businessType,
      country,
      hub: { city },
      region,
      skytraxRating,
    },
  } = airline ?? {}

  const localeStrings = useMemo(
    () => ({
      airplane: getLocale('airplane'),
      cargo: getLocale('cargo'),
      country: getLocale('country'),
      popularRoutes: getLocale('popularRoutes'),
      rating: getLocale('rating'),
      route: getLocale('route'),
    }),
    [selectedLocale],
  )

  const badge = useMemo(() => getAirlineBadge(businessType as AirlineType), [businessType])

  const locationText = useMemo(() => `${city}, ${country}, ${region}`, [city, country, region])

  const popularDestinations = useMemo(() => destinations?.slice(0, 4).join(', '), [destinations])

  const cargoLabel = useMemo(
    () =>
      businessModel === 'cargo' ? (
        <View className="bg-error px-2 py-1 rounded-xl overflow-hidden absolute top-2 left-2">
          <ThemedText color="text-100" type="button2">
            {localeStrings.cargo}
          </ThemedText>
        </View>
      ) : null,
    [businessModel, localeStrings],
  )

  const logAirlineCardPress = useCallback(async () => {
    await AnalyticsService.sendEvent('airline_card_press', {
      airline_id: id,
      airline_name: name,
      iata_code: iataCode,
      user_locale: selectedLocale,
    })
  }, [id, name, iataCode, selectedLocale])

  const handlePress = useCallback(() => {
    router.navigate({
      params: {
        airline: JSON.stringify(airline),
      },
      pathname: '/airline-detail',
    })
  }, [airline])

  const { showInterstitialAd } = useInterstitialAdHandler({
    adUnitId: AD_UNIT_ID,
  })

  const onCardPress = useCallback(() => {
    logAirlineCardPress()
    handlePress()
    showInterstitialAd()
  }, [logAirlineCardPress, handlePress, showInterstitialAd])

  return (
    <TouchableOpacity
      activeOpacity={0.7}
      className="bg-background-secondary rounded-xl mb-4 w-[48%] border border-background-quaternary shadow shadow-background-quaternary"
      hitSlop={20}
      onPress={onCardPress}
    >
      <View
        className="bg-background-primary rounded-t-xl overflow-hidden w-full justify-center"
        style={STATIC_STYLES.title}
      >
        <ThemedText color="tertiary-100" type="title" center>
          {iataCode}
        </ThemedText>

        <Image
          cachePolicy="memory-disk"
          contentFit="contain"
          source={badge}
          style={STATIC_STYLES.badge}
          transition={0}
        />

        <View className="bg-primary-100 px-2 py-1 rounded-xl overflow-hidden absolute bottom-2 right-2">
          <ThemedText color="text-100" type="button2">
            {iataCode}
          </ThemedText>
        </View>

        {cargoLabel}

        <FavoriteButton id={String(id)} type="airline" />
      </View>

      <View className="px-3 py-3">
        <View className="h-56 justify-between">
          <ThemedText
            color="text-100" ellipsizeMode="tail" numberOfLines={2}
            type="h4"
          >
            {name}
          </ThemedText>

          <ThemedText
            className="mt-1"
            color="text-100"
            ellipsizeMode="tail"
            numberOfLines={2}
            type="body2"
          >
            {locationText}
          </ThemedText>

          <View
            className={
              'bg-background-secondary rounded-xl p-2 shadow shadow-background-quaternary mt-2 border border-background-quaternary'
            }
          >
            <View className="flex-row items-center justify-between">
              <View className="flex-1 items-center">
                <ThemedText className=" mb-1" color="text-100" type="h4">
                  {destinationCount}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.route}
                </ThemedText>
              </View>

              <View className="w-[1px] h-6 bg-primary-100" />

              <View className="flex-1 items-center">
                <ThemedText className=" mb-1" color="text-100" type="h4">
                  {destinationCountries}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.country}
                </ThemedText>
              </View>
            </View>

            <View className="my-3 w-[100px] self-center flex flex-row justify-between">
              <View className="h-[1px] w-8 bg-primary-100" />
              <View className="h-[1px] w-8 bg-primary-100" />
            </View>

            <View className="flex-row items-center justify-between">
              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {totalAirplane}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.airplane}
                </ThemedText>
              </View>

              <View className="w-[1px] h-6 bg-primary-100" />
              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {skytraxRating}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.rating}
                </ThemedText>
              </View>
            </View>
          </View>
        </View>
      </View>

      <View className="h-20 px-3 mb-3">
        <ThemedText color="text-100" type="body3">
          {localeStrings.popularRoutes}
        </ThemedText>

        <View
          className="bg-background-quaternary px-2 py-1 rounded-xl overflow-hidden mt-2"
          style={STATIC_STYLES.containerWidth}
        >
          <View className="flex-row items-center px-1">
            <ThemedText
              color="text-90" ellipsizeMode="tail" numberOfLines={2}
              type="body4"
            >
              {popularDestinations}
            </ThemedText>
          </View>
        </View>
      </View>
    </TouchableOpacity>
  )
})

AirlineFavoriteItemCard.displayName = 'AirlineFavoriteItemCard'

export { AirlineFavoriteItemCard }
