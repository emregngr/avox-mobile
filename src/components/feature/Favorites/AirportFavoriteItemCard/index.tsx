import { Image } from 'expo-image'
import { router } from 'expo-router'
import React, { memo, useCallback, useMemo } from 'react'
import { Platform, TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { FavoriteButton } from '@/components/feature/FavoriteButton'
import { useInterstitialAdHandler } from '@/hooks/advertisement/useInterstitialAdHandler'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import type { AirportType } from '@/types/feature/airport'
import { AnalyticsService } from '@/utils/common/analyticsService'
import { responsive } from '@/utils/common/responsive'
import { getAirportImage } from '@/utils/feature/getAirportImage'
import type { AirportBadgeType } from '@/utils/feature/getBadge'
import { getAirportBadge } from '@/utils/feature/getBadge'

interface AirportFavoriteItemCardProps {
  airport: AirportType
}

const AD_UNIT_ID =
  Platform.OS === 'ios'
    ? 'ca-app-pub-4123130377375974/7531194946'
    : 'ca-app-pub-4123130377375974/8992450420'

const STATIC_STYLES = {
  badge: {
    height: 40,
    left: 4,
    position: 'absolute' as const,
    top: 4,
    width: 40,
  },
  containerWidth: {
    width: responsive.deviceWidth / 2 - 44,
  },
  image: {
    height: 130,
    width: '100%' as const,
  },
}

export const AirportFavoriteItemCard = memo(({ airport }: AirportFavoriteItemCardProps) => {
  const { selectedLocale } = useLocaleStore()

  const {
    facilities: { googleMapsRating },
    flightOperations: { airlines = [], destinationCount, destinationCountries },
    iataCode,
    icaoCode,
    id,
    image,
    infrastructure: { passengerCapacity },
    name,
    operations: {
      airportType,
      country,
      location: { city },
      region,
    },
  } = airport ?? {}

  const localeStrings = useMemo(
    () => ({
      country: getLocale('country'),
      passenger: getLocale('passenger'),
      popularAirlines: getLocale('popularAirlines'),
      rating: getLocale('rating'),
      route: getLocale('route'),
    }),
    [selectedLocale],
  )

  const defaultImage = useMemo(
    () => getAirportImage(airportType as AirportBadgeType),
    [airportType],
  )

  const badge = useMemo(() => getAirportBadge(airportType as AirportBadgeType), [airportType])

  const locationText = useMemo(() => `${city}, ${country}, ${region}`, [city, country, region])

  const airlinesText = useMemo(() => airlines?.map(airline => airline).join(', '), [airlines])

  const logAirportCardPress = useCallback(async () => {
    await AnalyticsService.sendEvent('airport_card_press', {
      airline_id: id,
      airline_name: name,
      iata_code: iataCode,
      user_locale: selectedLocale,
    })
  }, [id, name, iataCode, selectedLocale])

  const handlePress = useCallback(() => {
    router.navigate({
      params: {
        airport: JSON.stringify(airport),
      },
      pathname: '/airport-detail',
    })
  }, [airport])

  const { showInterstitialAd } = useInterstitialAdHandler({
    adUnitId: AD_UNIT_ID,
  })

  const onCardPress = useCallback(() => {
    logAirportCardPress()
    handlePress()
    showInterstitialAd()
  }, [logAirportCardPress, handlePress, showInterstitialAd])

  return (
    <TouchableOpacity
      activeOpacity={0.7}
      className="bg-background-secondary rounded-xl mb-4 w-[48%] border border-background-quaternary shadow shadow-background-quaternary"
      hitSlop={20}
      onPress={onCardPress}
      testID={`airport-card-${id}`}
    >
      <View className="rounded-t-xl overflow-hidden w-full justify-center">
        <Image
          cachePolicy="memory-disk"
          contentFit="cover"
          source={image || defaultImage}
          style={STATIC_STYLES.image}
          transition={0}
        />

        <Image
          cachePolicy="memory-disk"
          contentFit="contain"
          source={badge}
          style={STATIC_STYLES.badge}
          transition={0}
        />

        <View className="bg-primary-100 px-2 py-1 rounded-xl overflow-hidden absolute bottom-2 left-2">
          <ThemedText color="text-100" type="button2">
            {iataCode}
          </ThemedText>
        </View>

        <View className="bg-primary-100 px-2 py-1 rounded-xl overflow-hidden absolute bottom-2 right-2">
          <ThemedText color="text-100" type="button2">
            {icaoCode}
          </ThemedText>
        </View>

        <FavoriteButton id={id} type="airport" />
      </View>

      <View className="px-3 py-3">
        <View className="h-56 justify-between">
          <ThemedText
            color="text-100" ellipsizeMode="tail" numberOfLines={2}
            type="h4"
          >
            {name}
          </ThemedText>

          <ThemedText
            className="mt-1"
            color="text-90"
            ellipsizeMode="tail"
            numberOfLines={2}
            type="body2"
          >
            {locationText}
          </ThemedText>

          <View
            className={
              'bg-background-secondary rounded-xl p-2 shadow shadow-background-quaternary mt-2 border border-background-quaternary'
            }
          >
            <View className="flex-row items-center justify-between">
              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {destinationCount}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.route}
                </ThemedText>
              </View>

              <View className="w-[1px] h-6 bg-primary-100" />

              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {destinationCountries}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.country}
                </ThemedText>
              </View>
            </View>

            <View className="my-3 w-[100px] self-center -row justify-between">
              <View className="h-[1px] w-8 bg-primary-100" />
              <View className="h-[1px] w-8 bg-primary-100" />
            </View>

            <View className="flex-row items-center justify-between">
              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {passengerCapacity} m
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.passenger}
                </ThemedText>
              </View>

              <View className="w-[1px] h-6 bg-primary-100" />

              <View className="flex-1 items-center">
                <ThemedText className="mb-1" color="text-100" type="h4">
                  {googleMapsRating}
                </ThemedText>
                <ThemedText
                  color="text-90" lineBreakMode="tail" numberOfLines={1}
                  type="body4"
                >
                  {localeStrings.rating}
                </ThemedText>
              </View>
            </View>
          </View>
        </View>
      </View>

      <View className="h-20 px-3 mb-3">
        <ThemedText color="text-100" type="body3">
          {localeStrings.popularAirlines}
        </ThemedText>

        <View
          className="bg-background-quaternary px-2 py-1 rounded-xl overflow-hidden mt-2"
          style={STATIC_STYLES.containerWidth}
        >
          <View className="-row items-center px-1">
            <ThemedText
              color="text-90" ellipsizeMode="tail" numberOfLines={2}
              type="body4"
            >
              {airlinesText}
            </ThemedText>
          </View>
        </View>
      </View>
    </TouchableOpacity>
  )
})
