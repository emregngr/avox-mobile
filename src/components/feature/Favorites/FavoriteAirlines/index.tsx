import React, { useCallback, useMemo } from 'react'
import { FlatList, Platform, RefreshControl } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

import { AirlineFavoriteItemCard } from '@/components/feature/Favorites/AirlineFavoriteItemCard'
import { EmptyFavorite } from '@/components/feature/Favorites/EmptyFavorite'
import { FavoriteSkeleton } from '@/components/feature/Favorites/FavoriteSkeleton'
import { useBatchingPeriod } from '@/hooks/batchingPeriod/useBatchingPeriod'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { AirlineType } from '@/types/feature/airline'

interface FavoriteAirlinesListProps {
  airlines: AirlineType[]
  isLoading: boolean
  onRefresh: () => void
}

interface AirlineFavoriteItemCard {
  item: AirlineType
}

const INITIAL_ITEMS_PER_PAGE = 6
const MAX_ITEMS_PER_BATCH = 4
const NUM_COLUMNS = 2
const ITEM_HEIGHT = 500
const WINDOW_SIZE = 7

export const FavoriteAirlines = ({ airlines, isLoading, onRefresh }: FavoriteAirlinesListProps) => {
  const { bottom } = useSafeAreaInsets()

  const extraBottomPadding = Platform.OS === 'ios' ? 72 : 82
  const bottomPadding = bottom + extraBottomPadding

  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const BATCHING_PERIOD = useBatchingPeriod()

  const renderAirlineCard = useCallback(
    ({ item }: AirlineFavoriteItemCard) => <AirlineFavoriteItemCard airline={item} />,
    [],
  )

  const keyExtractor = useCallback((item: AirlineType) => item?.id, [])

  const getItemLayout = useCallback(
    (_: ArrayLike<AirlineType> | null | undefined, index: number) => ({
      index,
      length: ITEM_HEIGHT,
      offset: ITEM_HEIGHT * index,
    }),
    [],
  )

  if (isLoading) {
    return <FavoriteSkeleton type="airline" />
  }

  if (airlines?.length === 0) {
    return <EmptyFavorite icon="airplane" text={getLocale('noFavoriteAirline')} />
  }

  const refreshControl = (
    <RefreshControl
      colors={[colors?.tertiary100]}
      onRefresh={onRefresh}
      progressBackgroundColor={colors?.background?.primary}
      refreshing={isLoading}
      tintColor={colors?.tertiary100}
      title={getLocale('loading')}
      titleColor={colors?.text100}
    />
  )

  return (
    <FlatList
      className="mt-[52px]"
      columnWrapperClassName="justify-between"
      contentContainerClassName="px-4"
      contentContainerStyle={{ paddingBottom: bottomPadding }}
      data={airlines}
      getItemLayout={getItemLayout}
      initialNumToRender={INITIAL_ITEMS_PER_PAGE}
      keyExtractor={keyExtractor}
      maxToRenderPerBatch={MAX_ITEMS_PER_BATCH}
      numColumns={NUM_COLUMNS}
      refreshControl={refreshControl}
      renderItem={renderAirlineCard}
      scrollEventThrottle={16}
      showsVerticalScrollIndicator={false}
      updateCellsBatchingPeriod={BATCHING_PERIOD}
      windowSize={WINDOW_SIZE}
      removeClippedSubviews
    />
  )
}
