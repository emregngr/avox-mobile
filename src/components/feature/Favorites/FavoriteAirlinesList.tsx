import React, { useCallback, useMemo } from 'react'
import { FlatList, RefreshControl, View } from 'react-native'

import { AirlineFavoriteItemCard } from '@/components/feature/Airline/AirlineFavoriteItemCard'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { Airline } from '@/types/feature/airline'

import { EmptyState } from './EmptyState'
import { SkeletonList } from './SkeletonList'

const ITEMS_PER_PAGE = 6

interface FavoriteAirlinesListProps {
  airlines: Airline[]
  isLoading: boolean
  onRefresh: () => void
}

interface AirlineFavoriteItemCard {
  item: Airline
}

export const FavoriteAirlinesList = ({
  airlines,
  isLoading,
  onRefresh,
}: FavoriteAirlinesListProps) => {
  const { selectedTheme } = useThemeStore()
  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const renderAirlineCard = useCallback(
    ({ item }: AirlineFavoriteItemCard) => <AirlineFavoriteItemCard airline={item} />,
    [],
  )

  if (isLoading) {
    return <SkeletonList type="airline" />
  }

  if (airlines?.length === 0) {
    return (
      <View className="flex-1 mt-14 px-4 pb-10">
        <EmptyState text={getLocale('noFavoriteAirline')} />
      </View>
    )
  }

  const refreshControl = (
    <RefreshControl
      colors={[colors?.tertiary100]}
      onRefresh={onRefresh}
      progressBackgroundColor={colors?.background?.primary}
      refreshing={isLoading}
      tintColor={colors?.tertiary100}
      title={getLocale('loading')}
      titleColor={colors?.text100}
    />
  )

  return (
    <FlatList
      className="mt-[52px]"
      columnWrapperClassName="justify-between"
      contentContainerClassName="px-4 pb-10"
      data={airlines}
      initialNumToRender={ITEMS_PER_PAGE}
      keyExtractor={(item, index) => `airline-${item?.id || index}`}
      maxToRenderPerBatch={ITEMS_PER_PAGE}
      numColumns={2}
      refreshControl={refreshControl}
      renderItem={renderAirlineCard}
      showsVerticalScrollIndicator={false}
      updateCellsBatchingPeriod={200}
      windowSize={5}
      removeClippedSubviews
    />
  )
}
