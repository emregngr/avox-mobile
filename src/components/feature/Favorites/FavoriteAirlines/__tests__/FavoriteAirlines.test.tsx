import { render } from '@testing-library/react-native'
import React from 'react'

import { FavoriteAirlines } from '@/components/feature/Favorites/FavoriteAirlines'
import useThemeStore from '@/store/theme'
import type { AirlineType } from '@/types/feature/airline'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('@/components/feature/Favorites/FavoriteSkeleton', () => {
  const { Text } = require('react-native')

  return {
    FavoriteSkeleton: ({ type }: { type: string }) => <Text>Skeleton {type}</Text>,
  }
})

jest.mock('@/components/feature/Favorites/EmptyFavorite', () => {
  const { Text } = require('react-native')

  return {
    EmptyFavorite: ({ text }: { text: string }) => <Text>Empty {text}</Text>,
  }
})

jest.mock('@/components/feature/Favorites/AirlineFavoriteItemCard', () => {
  const { Text } = require('react-native')

  return {
    AirlineFavoriteItemCard: ({ airline }: { airline: AirlineType }) => (
      <Text>{`AirlineCard ${airline.name}`}</Text>
    ),
  }
})

const mockedAirlines: any = [
  { id: '1', name: 'THY' },
  { id: '2', name: 'Pegasus' },
]

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })
})

describe('FavoriteAirlines Component', () => {
  it('renders skeleton when loading', () => {
    const { getByText } = render(<FavoriteAirlines airlines={[]} onRefresh={jest.fn()} isLoading />)
    expect(getByText('Skeleton airline')).toBeTruthy()
  })

  it('renders empty state when no airlines', () => {
    const { getByText } = render(
      <FavoriteAirlines airlines={[]} isLoading={false} onRefresh={jest.fn()} />,
    )
    expect(getByText('Empty noFavoriteAirline')).toBeTruthy()
  })

  it('renders airline cards when data is provided', () => {
    const { getByText } = render(
      <FavoriteAirlines airlines={mockedAirlines} isLoading={false} onRefresh={jest.fn()} />,
    )
    expect(getByText('AirlineCard THY')).toBeTruthy()
    expect(getByText('AirlineCard Pegasus')).toBeTruthy()
  })
})

describe('FavoriteAirlines Component Snapshot', () => {
  it('should render the FavoriteAirlines Component successfully', () => {
    const { toJSON } = render(<FavoriteAirlines airlines={[]} onRefresh={jest.fn()} isLoading />)

    expect(toJSON()).toMatchSnapshot()
  })
})
