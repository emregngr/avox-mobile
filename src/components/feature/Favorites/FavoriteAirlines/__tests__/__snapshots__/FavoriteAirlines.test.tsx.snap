// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FavoriteAirlines Component Snapshot should render the FavoriteAirlines Component successfully 1`] = `
<Text>
  Skeleton 
  airline
</Text>
`;
